### PR TITLE
Middleware for request / success / failure in single action with some API handling

### PR DESCRIPTION
**Pull request by AriHrannar**
## Status
**Ready**

## Description
Creating middleware that allows us to write actions much clearer (see get politician action)

### CODE REVIEW

File: src/Components/Promises/PromiseForm/index.js [hunk 1]
```diff
@@ -65,6 +65,7 @@ export class PromiseForm extends React.Component {
 
   render() {
     const { parties, politicians } = this.props;
+    if (!politicians) return null;
```
Comment (Ragnar-H) [inline src/Components/Promises/PromiseForm/index.js]: can we get rid of the wrapping `div` from the `render` function?

File: src/Stores/Politicians/actions.js [hunk 2]
```diff
@@ -1,31 +1,17 @@
...
-});
-
-export const fetchPoliticians = () => (dispatch: Dispatch) => {
-  dispatch(getPoliticians());
-
-  return apiGetPoliticians().then(
-    resp => resp.data && dispatch(getPoliticiansSuccess(resp.data)),
-    error => dispatch(getPoliticiansFailure(error.message))
-  );
+export const fetchPoliticians = () => {
```
Comment (Ragnar-H) [inline src/Stores/Politicians/actions.js]: ❤️ this

File: src/middleware/api.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+};
+
+export default () => dispatch => async action => {
+  const { endpoint, method, types, content } = action;
+  if (typeof types === 'undefined') {
+    // So the middleware doesn't get applied to every single action
+    return dispatch(action);
+  }
+
+  const requestType = types.request;
```
Comment (Ragnar-H) [inline src/middleware/api.js]: nitpick:
maybe a destructure call is easier to read here

File: src/middleware/api.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  const payload = await response.json();
+
+  if (!ok) {
+    if (status === STATUS_CODES.UNAUTHORIZED) {
+      // redirect to login
+    }
+
+    return dispatch({
+      errors: payload && payload.errors ? payload.errors : [],
```
Comment (Ragnar-H) [inline src/middleware/api.js]: Probably makes sense to have a fallback error, defined in a constant.

`const UNHANDLED_ERROR = { message: 'oh no gooses!' }`

File: src/Components/Promises/PromiseForm/index.js (hunk 1 above)
Comment (AriHrannar) [reply src/Components/Promises/PromiseForm/index.js]: yes that sounds good!

File: src/Stores/Politicians/actions.js (hunk 2 above)
Comment (AriHrannar) [reply src/Stores/Politicians/actions.js]: so efficient

File: src/middleware/api.js (hunk 3 above)
Comment (AriHrannar) [reply src/middleware/api.js]: good suggestion, adding it

File: src/middleware/api.js (hunk 4 above)
Comment (AriHrannar) [reply src/middleware/api.js]: Good suggestion, this has been added

**Pull request merged**